sklearn/preprocessing/_data.py

```diff
@@ -1802,6 +1806,342 @@ def transform(self, X):
+        dist_min = base_knots[1] - base_knots[0]
+        dist_max = base_knots[-1] - base_knots[-2]
+        knots = np.r_[linspace(base_knots[0] - degree * dist_min,
+                               base_knots[0] - dist_min, num=degree),
+                      base_knots,
+                      linspace(base_knots[-1] + dist_max,
+                               base_knots[-1] + degree * dist_max,
+                               num=degree)]
+
+        # With a diagonal coefficient matrix, we get back the spline basis
+        # elements, i.e. the design matrix of the spline.
+        # Note, BSpline appreciate C-contiguous arrays.
+        coef = np.eye(n_knots + self.degree - 1, dtype=np.float64)
+        extrapolate = self.extrapolation == "continue"
+        bsplines = [BSpline(knots[:, i], coef, self.degree,
+                            extrapolate=extrapolate)
+                    for i in range(n_features)]
+        self.bsplines_ = np.array(bsplines)
+
+        self.n_input_features_ = n_features
+        self.n_output_features_ = n_out - n_features * self.include_bias
+        return self
+
+    def transform(self, X):
+        """Transform each feature data to B-splines.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data to transform.
+
+        Returns
+        -------
+        XBS : ndarray of shape (n_samples, n_features * n_splines)
+            The matrix of features, where n_splines is the number of bases
+            elements of the B-splines, n_knots + degree - 1.
+        """
+        check_is_fitted(self)
+
+        X = check_array(X)
+
+        n_samples, n_features = X.shape
+        n_splines = self.bsplines_[0].c.shape[0]
+        degree = self.degree
+
+        if n_features != self.n_input_features_:
+            raise ValueError("X shape does not match training shape")
+
+        # Note that scipy splines return float64.
+        n_out = self.n_output_features_ + n_features * self.include_bias
+        XBS = np.zeros((n_samples, n_out), dtype=np.float64,
+                       order=self.order)
+
+        for i in range(n_features):
+            spl = self.bsplines_[i]
+
+            if self.extrapolation in ('continue', 'error'):
+                XBS[:, (i * n_splines):((i + 1) * n_splines)] = spl(X[:, i])
+            else:
+                xmin = spl.t[degree]
+                xmax = spl.t[-degree - 1]
+                mask = (xmin <= X[:, i]) & (X[:, i] <= xmax)
+                XBS[mask, (i * n_splines):((i + 1) * n_splines)] = (
+                    spl(X[mask, i]))
+
+            # Note for extrapolation:
+            # 'continue' is already returned as is by scipy BSplines
+            if self.extrapolation == 'error':
+                # BSpline with extrapolate=False does not raise an error, but
+                # output np.nan.
+                assert_all_finite(
+                    XBS[:, (i * n_splines):((i + 1) * n_splines)])
```

Comment:
Usually `assert_all_finite` gives an error message that a user would not be able to interpret:

```py
if np.isnan(...):
    raise ValueError("X contains values beyond the limits of the knots")
```

or something similar.